### PR TITLE
Create Counter model to collect information for menu counts.

### PR DESCRIPTION
**Pull request by alex-mtl**
New UserAPI endpoint implemented to get all menu counts:
`{{host}}/api/v2/users/me-counts`

New interface `CounterProviderInterface` implemented in:

- `ActivityCounterProvider` (return: **UnreadNotifications**)
- `ParticipatedCounterProvider` (return: **Participated**) https://github.com/vanilla/addons/pull/697
- `LogCounterProvider` (return: **SpamQueue** and **ModerationQueue**)
- `RoleCounterProvider` (return: **Applicants**)
- `UserCounterProvider` (return: **Bookmarks**, **Discussions**, **UnreadDiscussions**, **Conversations**, **Drafts**)

Openapi specs are updated: 

- `applications/dashboard/openapi/users.yml`

### CODE REVIEW

File: applications/vanilla/settings/class.hooks.php [hunk 1]
```diff
@@ -25,6 +25,14 @@ public function container_init(Container $dic) {
         $dic->rule(\Vanilla\Navigation\BreadcrumbModel::class)
             ->addCall('addProvider', [new Reference(\Vanilla\Forum\Navigation\ForumBreadcrumbProvider::class)])
         ;
+
+        $dic->rule(\Vanilla\Menu\CounterModel::class)
+            ->addCall('addProvider', [new Reference(\Vanilla\Forum\Menu\DiscussionCounterProvider::class)])
+            ->addCall('addProvider', [new Reference(\Vanilla\Forum\Menu\UserCounterProvider::class)])
+            ->addCall('addProvider', [new Reference(\Vanilla\Forum\Menu\ActivityCounterProvider::class)])
+            ->addCall('addProvider', [new Reference(\Vanilla\Forum\Menu\LogCounterProvider::class)])
+            ->addCall('addProvider', [new Reference(\Vanilla\Forum\Menu\RoleCounterProvider::class)])
```
[the remark this answers is not in the document]
Comment (alex-mtl) [reply applications/vanilla/settings/class.hooks.php]: We have all our core hooks here under vanilla application.
Am I wrong? do we have any hook file for dashboard?

[the remark this answers is not in the document]
Comment (initvector) [reply applications/vanilla/settings/class.hooks.php]: > do we have any hook file for dashboard?

Yep! Dashboard has its own hooks file. There's already a [`container_init` hook](https://github.com/vanilla/vanilla/blob/d84a3b8862a418728df0b2acd3194d4b1819716f/applications/dashboard/settings/class.hooks.php#L50) in there too.

File: applications/vanilla/library/MenuCounts/DiscussionCounterProvider.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public function __construct(\DiscussionModel $discussionModel) {
+        $this->discussionModel = $discussionModel;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getMenuCounters(): array {
+        $counters[] = new Counter("Participated", $this->discussionModel->getCountParticipated());
```
Comment (initvector) [inline applications/vanilla/library/MenuCounts/DiscussionCounterProvider.php]: Getting the following fatal error from this line.

> The "DiscussionModel" object does not have a "xgetCountParticipated" method.

Comment (alex-mtl) [reply applications/vanilla/library/MenuCounts/DiscussionCounterProvider.php]: Good catch. Moved to addons repo. Please check: https://github.com/vanilla/addons/pull/697

File: applications/vanilla/settings/class.hooks.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (alex-mtl) [reply applications/vanilla/settings/class.hooks.php]: Splited.

File: tests/APIv2/UsersTest.php [hunk 3]
```diff
@@ -204,6 +204,18 @@ public function testMeGuest() {
...
+    /**
+     * Test getting current menu counts /users/me-counts.
+     */
+    public function testMeCounts() {
+        $response = $this->api()->get("{$this->baseUrl}/me-counts");
+        $this->assertSame(200, $response->getStatusCode());
+
+        $response = $response->getBody();
+
+        $this->assertArrayHasKey('counts', $response);
```
[the remark this answers is not in the document]
Comment (alex-mtl) [reply tests/APIv2/UsersTest.php]: Can we create separate issue for testing? 
We need to cover `vanilla/addons/plugins/participated` and `vanilla/internal/plugins/badges` as well and etc.

[the remark this answers is not in the document]
Comment (initvector) [reply tests/APIv2/UsersTest.php]: I think that's reasonable. Vanilla might have some bootstrapping issues in its tests that might also need to be addressed. Will you please create the issue and crosslink it here?

**Comment by initvector** [comment]
I think this is just about done, but there's a couple small things lingering:

1. The activity, log and role count provider hooks were moved into core, which is great, but the classes are still living under the Vanilla addon. Can we move those into Dashboard?
1. Conversations' count provider and its hook should be moved out under the Conversations addon, similar to what you did with Dashboard. Otherwise, if you disable the addon, you'll still see the counts.

**Pull request merged**